src/openms/source/ANALYSIS/QUANTITATION/AbsoluteQuantitation.cpp

```diff
@@ -0,0 +1,438 @@
+    Feature empty_feature;
+    size_t IS_component_it, IS_component_group_it;
+
+    // // iterate through the unknowns
+    // for (size_t i = 0; i < unknowns.size(); i++)
+    // {      
+
+    // iterate through each component_group/feature     
+    for (size_t feature_it = 0; feature_it < unknowns.size(); ++feature_it)
+    {
+      component_group_name = (String)unknowns[feature_it].getMetaValue("PeptideRef");
+      Feature unknowns_quant_feature;
+
+      // iterate through each component/sub-feature
+      for (size_t sub_it = 0; sub_it < unknowns[feature_it].getSubordinates().size(); ++sub_it)
+      {
+        component_name = (String)unknowns[feature_it].getSubordinates()[sub_it].getMetaValue("native_id");  
+
+        // apply the calibration curve to components that are in the quant_method
+        if (quant_methods_.count(component_name)>0)
+        { 
+          calculated_concentration = 0.0;    
+          quant_methods_it = quant_methods_.find(component_name);
+          quant_methods_it->second.getComponentISFeatureNames(quant_component_name,quant_IS_component_name,quant_feature_name);
+          if (quant_IS_component_name != "")
+          {
+            // look up the internal standard for the component
+            IS_found = false;
+            // Optimization: 90% of the IS will be in the same component_group/feature
+            for (size_t is_sub_it = 0; is_sub_it < unknowns[feature_it].getSubordinates().size(); ++is_sub_it)
+            {
+              IS_component_name = (String)unknowns[feature_it].getSubordinates()[is_sub_it].getMetaValue("native_id");              
+              if (quant_IS_component_name == IS_component_name)
+              {
+                IS_found = true;
+                IS_component_group_it = feature_it;
+                IS_component_it = is_sub_it;
+                break;
+              }
+            }
+            if (!IS_found)
+            {// expand IS search to all components                
+              // iterate through each component_group/feature     
+              for (size_t is_feature_it = 0; is_feature_it < unknowns.size(); ++is_feature_it)
+              {
+                //iterate through each component/sub-feature
+                for (size_t is_sub_it = 0; is_sub_it < unknowns[is_feature_it].getSubordinates().size(); ++is_sub_it)
+                {
+                  IS_component_name = (String)unknowns[is_feature_it].getSubordinates()[is_sub_it].getMetaValue("native_id");                   
+                  if (quant_IS_component_name == IS_component_name)
+                  {
+                    IS_found = true;
+                    IS_component_group_it = is_feature_it;
+                    IS_component_it = is_sub_it;
+                    break;
+                  }
+                }
+                if (IS_found)
+                {
+                  break;
+                }
+              }
+            }
+            if (IS_found)
```

Comment:
you could factor out the whole method searching for the IS into its own fxn, e.g. lines 210-250

Reply:
Good idea.